### PR TITLE
Closes #699: Automatic deploys to staging on commit to master

### LINKED ISSUE #699: Automatic deploys to staging on commit to master
**What would you like to be added**:

Let's add a webhook flow for our staging deployment builds.  When we commit on the `master` branch of the  Seneca-CDOT/telescope repo, we should trigger a build process on dev.telescope.cdot.systems.

I think the way to do this is using a [GitHub webhook] for the [push event](https://developer.github.com/v3/activity/events/types/#pushevent) from our repo on GitHub.

For this to work, we'll need to have a POST route running on our telescope server (maybe at `/admin/webhooks/push`).  There are various pre-made middleware routes for this, for example [github-webhook-handler](https://www.npmjs.com/package/github-webhook-handler). We'll also need to [secure the telescope side of the webhook on our server](https://developer.github.com/webhooks/securing/), so we don't respond to push events from anyone but GitHub.

In the payload we receive, we'll have to look for `ref` to be `refs/head/master` (i.e., for a push to the master branch).  We could also do a bit of sanity checking, to get the current commit SHA and the previous commit SHA so we can compare with what's on the server's git repo (could do this in follow-up).

Then, we need to trigger an update to our git repo on the telescope server and kick off a docker rebuild.  It could be as simple as this:

```bash
#!/bin/bash
git pull origin master
docker-compose up -d
```

But I suspect @manekenpix has thoughts about what he's been doing manually that we can turn into a script that can live in our `tools/` directory.

**Why would you like this to be added**:

I want to move us toward "Continuous Delivery" and limit the amount of work we expect @manekenpix to do for us manually.  He's done an amazing job, but we can't expect him to always handle deployment.

### PR DESCRIPTION
**Pull request by manekenpix**
## Issue This PR Addresses
Closes #699 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
Added automatic deployment to staging on commit to master.
Using webhooks, merges to master will trigger automatic redeployment to include latest changes to Telescope.

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: env.staging [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# NODE_ENV should be one of "development" or "production"
+NODE_ENV=development
```
Comment (humphd) [inline env.staging]: I think this should be `production` so it's closer to what we'll do on prod.

File: env.staging [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+# API_URL is the URL of the Telescope Web API server.  Change this to
+# pick which backend server our frontend uses.  If you are developing
+# locally, this will be localhost:{PORT}, probably http://localhost:3000.
+# Other possible values include:
+#
+#   API_URL=http://localhost:3000 (local development)
+API_URL=https://dev.telescope.cdot.systems
+#   API_URL=https://telescope.cdot.systems (our production server)
+#   API_URL=http://localhost:3000
```
Comment (humphd) [inline env.staging]: I would delete all the lines that you aren't using.  In development it's nice to have them around, but in this case, let's strip this file down to the bare essentials, and not leave any extra stuff in here we could mistakenly enable.

File: tools/autodeployment/server.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+const { SECRET, REPO_NAME, DEPLOY_PORT } = process.env;
+
+const handler = createHandler({ path: '/', secret: SECRET });
+
+http.createServer((req, res) => {
+  handler(req, res, err => {
+    res.statusCode = 404
+    res.end('Page not found');
+  })
+}).listen(CD_PORT, () => {
```
Comment (humphd) [inline tools/autodeployment/server.js]: What is CD_PORT?

File: tools/autodeployment/deploy_stage [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (humphd) [inline tools/autodeployment/deploy_stage]: Can you name this file `.sh` so it's more clear what it is?

File: tools/autodeployment/server.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    console.log(`Server listening on port ${DEPLOY_PORT}`);
```
Comment (humphd) [inline tools/autodeployment/server.js]: We could do this later, but at some point we probably need to figure out how to get this to log somewhere we can go read what happened when things fail.

File: tools/autodeployment/package.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "autodeployment",
+  "version": "0.0.1",
+  "description": "A tool for automatic deployment",
+  "main": "server.js",
+  "dependencies": {
+    "dotenv": "^8.2.0",
+    "github-webhook-handler": "^1.0.0",
+    "shelljs": "^0.8.3"
```
Comment (humphd) [inline tools/autodeployment/package.json]: You had talked about using pm2.  Are you going to do that in a different PR?

Comment (manekenpix) [reply tools/autodeployment/package.json]: Added pm2

**Comment by manekenpix** [comment]
For documentation, I'll file an issue once it gets merged and I'll assign it to myself and @c3ho.

File: tools/autodeployment/deploy_stage.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+docker system prune -af
+
+
+# Delete and clone
+cd ..
+
+rm -rf telescope
+
+git clone https://github.com/Seneca-CDOT/telescope.git
```
Comment (humphd) [inline tools/autodeployment/deploy_stage.sh]: An improvement we could make in a future PR is to checkout the specific commit sha that was just pushed vs. whatever is currently on `master`.  It's unlikely they will be different, but if they are, we could have odd bugs.

File: tools/autodeployment/env.example [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# Port on which the server will be listening
+DEPLOY_PORT=
+
+# Secret to be shared between GitHub and the server
+SECRET=
+
+# Repo that will be automatically deployed
```
Comment (humphd) [inline tools/autodeployment/env.example]: Can you give an example of the format you want for this?  What is a "repo name"? (with .git? without? as a URL?)

File: tools/autodeployment/package.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "autodeployment",
+  "version": "0.0.1",
+  "description": "A tool for automatic deployment",
+  "main": "server.js",
+  "dependencies": {
+    "dotenv": "^8.2.0",
+    "github-webhook-handler": "^1.0.0",
+    "pm2": "^4.2.3",
```
Comment (humphd) [inline tools/autodeployment/package.json]: I started to type out changes to do based on adding this, and realized it's too big.  Let's follow-up with pm2 changes later.

File: tools/autodeployment/server.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  });
+
+handler.on('error', err => {
+  console.error('Error:', err.message);
+});
+
+handler.on('push', event => {
+  const { name } = event.payload.repository;
+
+  if (name === REPO_NAME) {
```
Comment (humphd) [inline tools/autodeployment/server.js]: In a follow-up PR, can we get the commit sha here and pass it to the script?
